File: leetcode/leetcode/050.cpp
//
//  050.cpp
//  leetcode
//
//  Created by Hang Zhang on 9/11/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
#include <map>
#include <limits>
using namespace std;

class Solution {
public:
    double myPow(double x, int n) {
        if (n == 0)
            return 1;
        if (n < 0)
            x = 1/x;
        double result = 1;
        unsigned int tn = n > 0 ? n : -n;
        while (tn > 0) {
            result *= tn&1 ? x : 1;
            x *= x;
            tn = tn >> 1;
        }
        return result;
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 050 Pow(x, n), C++ ...\n\n";

    double x = 34.00515;
    int n = -3;

    Solution sol;
    double result = sol.myPow(x, n);
    cout << x << " ^ " << n << " = " << result;

    return 0;
}

File: python/050.py
#   050
#   Pow(x, n)
#   2015-09-14
#####################################################
class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n == 0:
            return 1
        if n < 0:
            x = 1/x
            n = -n
        result = 1
        while n > 0:
            result *= x if n&1>0 else 1
            x *= x
            n = n>>1
        return result

#####################################################

sol = Solution()
z = sol.myPow(2.5, 2)
print(z)

